libraries/core_libs/consensus/src/pbft/pbft_manager.cpp

```diff
@@ -1430,23 +1431,45 @@ std::shared_ptr<PbftBlock> PbftManager::identifyLeaderBlock_(uint64_t round, uin
       continue;
     }
 
-    leader_candidates[getProposal(v)] = v;
+    if (pbft_chain_->findPbftBlockInChain(proposed_block_hash)) {
+      continue;
+    }
+
+    auto block = proposed_blocks_.getPbftProposedBlock(v->getPeriod(), v->getRound(), proposed_block_hash);
+    if (!block) {
+      LOG(log_er_) << "Unable to find proposed block " << proposed_block_hash;
+      continue;
+    }
+
+    if (block->getPivotDagBlockHash() == NULL_BLOCK_HASH) {
+      leader_candidates_empty_blocks[getProposal(v)] = block;
+    } else {
+      leader_candidates[getProposal(v)] = block;
+    }
   }
 
-  for (auto const &leader_vote : leader_candidates) {
-    const auto proposed_block_hash = leader_vote.second->getBlockHash();
+  for (auto const &leader : leader_candidates) {
```

Comment:
if we wanted to be fancy and not repeat code, this could be in lambda function:

```
  // Empty blocks are only selected as leader if there is no block proposed with dag blocks
  for (auto const &leader : leader_candidates_empty_blocks) {
    auto leader_block = leader.second;

    if (!compareBlocksAndRewardVotes_(leader_block)) {
      LOG(log_er_) << "Incomplete or invalid proposed block " << leader_block->getBlockHash() << ", period " << period
                   << ", round " << round;
      continue;
    }
    if (!pbft_chain_->checkPbftBlockValidation(leader_block)) {
      LOG(log_er_) << "Proposed block " << leader_block->getBlockHash() << " failed validation, period " << period
                   << ", round " << round;
      continue;
    }
    return leader_block;
  }

```


and we just call it with leader_candidates and leader_candidates_empty_blocks

Reply:
Done